lib/handler/traffic.c

```diff
@@ -0,0 +1,144 @@
+#include <stdlib.h>
+#include "h2o.h"
+
+#ifndef HUNDRED_MS
+#define HUNDRED_MS 100
+#endif
+
+#ifndef ONE_SECOND
+#define ONE_SECOND 1000
+#endif
+
+typedef H2O_VECTOR(h2o_iovec_t) iovec_vector_t;
+
+typedef struct st_traffic_shaper_t {
+    h2o_ostream_t super;
+    h2o_timeout_entry_t timeout_entry;
+    size_t tokens;
+    size_t token_inc;
+    h2o_context_t *ctx;
+    h2o_req_t *req;
+    struct {
+        iovec_vector_t bufs;
+        size_t inbufcnt;
+        int is_final;
+    } state;
+} traffic_shaper_t;
+
+static int real_send(traffic_shaper_t *self) {
+    /* a really simple token bucket implementation */
+    assert(self->tokens);
+    size_t i, token_consume;
+
+    token_consume = 0;
+
+    for (i = 0; i < self->state.inbufcnt; i++) {
+        if (self->state.bufs.entries[i].len > self->tokens - token_consume) {
+            return 0;
+        }
```

Comment:
Let's change `tokens` to a signed value (possibly `ssize_t`), and send immediately if current value of `tokens` is greater than zero.  I believe it would be safer to first send some data and then throttle, rather than to first throttle and then send.